File: src/test/scala/org/apache/flink/streaming/connectors/pulsar/SchemaITest.scala
/**
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.flink.streaming.connectors.pulsar

import java.util.Date
import java.util.Properties
import java.util.concurrent.atomic.AtomicInteger
import java.sql.Timestamp
import java.text.SimpleDateFormat
import java.sql.{Date => sDate}

import org.apache.flink.api.common.typeinfo.TypeInformation
import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment
import org.apache.flink.streaming.connectors.pulsar.internals.{PulsarFlinkTest, PulsarFunSuite}
import org.apache.flink.streaming.connectors.pulsar.testutils.FailingIdentityMapper
import org.apache.flink.table.descriptors.{Pulsar, Schema}
import org.apache.flink.table.runtime.utils.StreamITCase
import org.apache.flink.table.types.DataType
import org.apache.flink.table.types.utils.LegacyTypeInfoDataTypeConverter
import org.apache.flink.types.Row
import org.apache.flink.util.StringUtils
import org.apache.pulsar.common.naming.TopicName
import org.apache.pulsar.common.schema.SchemaType

import scala.reflect.ClassTag

class SchemaITest extends PulsarFunSuite with PulsarFlinkTest {

  import org.apache.flink.streaming.connectors.pulsar.internals.SchemaData._
  import org.apache.flink.streaming.connectors.pulsar.internal.PulsarOptions._
  import org.apache.flink.table.api._
  import org.apache.flink.table.api.scala._

  override def beforeEach(): Unit = {
    super.beforeEach()
    StreamITCase.testResults.clear()
    FailingIdentityMapper.failedBefore = false
  }

  test("test boolean read") {
    checkRead[Boolean](SchemaType.BOOLEAN, booleanSeq, null)
  }

  test("test boolean write") {
    checkWrite[Boolean](SchemaType.BOOLEAN, DataTypes.BOOLEAN(), booleanSeq, null)
  }

  test("test int read") {
    checkRead[Int](SchemaType.INT32, int32Seq, null)
  }

  test("test int write") {
    checkWrite[Int](SchemaType.INT32, DataTypes.INT(), int32Seq, null)
  }

  test("test string read") {
    checkRead[String](SchemaType.STRING, stringSeq, null)
  }

  test("test string write") {
    checkWrite[String](SchemaType.STRING, DataTypes.STRING(), stringSeq, null)
  }

  test("test byte read") {
    checkRead[Byte](SchemaType.INT8, int8Seq, null)
  }

  test("test byte write") {
    checkWrite[Byte](SchemaType.INT8, DataTypes.TINYINT(), int8Seq, null)
  }

  test("test short read") {
    checkRead[Short](SchemaType.INT16, int16Seq, null)
  }

  test("test short write") {
    checkWrite[Short](SchemaType.INT16, DataTypes.SMALLINT(), int16Seq, null)
  }

  test("test float read") {
    checkRead[Float](SchemaType.FLOAT, floatSeq, null)
  }

  test("test float write") {
    checkWrite[Float](SchemaType.FLOAT, DataTypes.FLOAT(), floatSeq, null)
  }

  test("test double read") {
    checkRead[Double](SchemaType.DOUBLE, doubleSeq, null)
  }

  test("test double write") {
    checkWrite[Double](SchemaType.DOUBLE, DataTypes.DOUBLE(), doubleSeq, null)
  }

  test("test date read") {
    val dateFormat = new SimpleDateFormat("yyyy-MM-dd")
    checkRead[Date](SchemaType.DATE, dateSeq, dateFormat.format(_))
  }

  test("test date write") {
    val dateFormat = new SimpleDateFormat("yyyy-MM-dd")
    checkWrite[sDate](SchemaType.DATE, DataTypes.DATE().bridgedTo(classOf[sDate]),
      dateSeq.map(d => new sDate(d.getTime))
      , dateFormat.format(_))
  }

  test("test timestamp read") {
    checkRead[Timestamp](SchemaType.TIMESTAMP, timestampSeq, null)
  }

  test("test timstamp write") {
    checkWrite[Timestamp](SchemaType.TIMESTAMP, DataTypes.TIMESTAMP(3).bridgedTo(classOf[Timestamp]), timestampSeq, null)
  }

  test("test byte array read") {
    checkRead[Array[Byte]](SchemaType.BYTES, bytesSeq, StringUtils.arrayAwareToString)
  }

  test("test byte array write") {
    checkWrite[Array[Byte]](SchemaType.BYTES, DataTypes.BYTES(), bytesSeq, StringUtils.arrayAwareToString)
  }

  private def checkRead[T: ClassTag](
      schemaType: SchemaType,
      datas: Seq[T],
      str: T => String) = {

    val see = StreamExecutionEnvironment.getExecutionEnvironment
    see.setParallelism(1)
    val tEnv = StreamTableEnvironment.create(see)

    val table = newTopic()
    sendTypedMessages[T](table, schemaType, datas, None)

    val props = sourceProperties()
    props.setProperty(TOPIC_SINGLE, table)

    tEnv
      .connect(new Pulsar().properties(props))
      .inAppendMode()
      .registerTableSource(table)

    val t: Table = tEnv.scan(table).select("value")
    implicit val ti = t.getSchema.toRowType
    val as = t.toAppendStream[Row]
    as.map(new FailingIdentityMapper[Row](datas.length))

    as.addSink(new StreamITCase.StringSink[Row]).setParallelism(1)

    val runner = new Thread("read") {
      override def run(): Unit = {
        try {
          see.execute("read")
        } catch {
          case t: Throwable => // do nothing
        }
      }
    }
    runner.start()
    runner.join()

    if (str == null) {
      assert(StreamITCase.testResults == datas.init.map(_.toString))
    } else {
      assert(StreamITCase.testResults == datas.init.map(str(_)))
    }
  }

  private def checkWrite[T: ClassTag](
    schemaType: SchemaType,
    dt: DataType,
    datas: Seq[T],
    str: T => String) = {

    val see = StreamExecutionEnvironment.getExecutionEnvironment
    see.setParallelism(1)
    val tEnv = StreamTableEnvironment.create(see)

    val table = newTopic()
    val tableName = TopicName.get(table).getLocalName

    val tSchema = TableSchema.builder().field("value", dt).build()

    val stream =
      see.fromCollection(datas)(LegacyTypeInfoDataTypeConverter.toLegacyTypeInfo(dt).asInstanceOf[TypeInformation[T]])
    tEnv.registerDataStream("origin", stream)

    val sinkProps = sinkProperties()
    sinkProps.setProperty(TOPIC_SINGLE, table)

    tEnv
      .connect(new Pulsar().properties(sinkProps))
      .withSchema(new Schema().schema(tSchema))
      .inAppendMode()
      .registerTableSink(tableName)

    tEnv.sqlUpdate(s"insert into `$tableName` select * from origin")

    val sinkThread = new Thread("sink") {
      override def run(): Unit = {
        try {
          see.execute("sink")
        } catch {
          case t: Throwable => // do nothing
        }
      }
    }
    sinkThread.start()
    sinkThread.join()

    val se2 = StreamExecutionEnvironment.getExecutionEnvironment
    se2.setParallelism(1)
    val tEnv2 = StreamTableEnvironment.create(se2)

    val props = sourceProperties()
    props.setProperty(TOPIC_SINGLE, table)

    tEnv2
      .connect(new Pulsar().properties(props))
      .inAppendMode()
      .registerTableSource(table)

    val t: Table = tEnv2.scan(table).select("value")
    implicit val ti = t.getSchema.toRowType
    val as = t.toAppendStream[Row]
    as.map(new FailingIdentityMapper[Row](datas.length))

    as.addSink(new StreamITCase.StringSink[Row]).setParallelism(1)

    val runner = new Thread("read") {
      override def run(): Unit = {
        try {
          se2.execute("read")
        } catch {
          case t: Throwable => // do nothing
        }
      }
    }
    runner.start()
    runner.join()

    if (str == null) {
      assert(StreamITCase.testResults == datas.init.map(_.toString))
    } else {
      assert(StreamITCase.testResults == datas.init.map(str(_)))
    }
  }

  def sourceProperties(): Properties = {
    val prop = new Properties()
    prop.setProperty(SERVICE_URL_OPTION_KEY, serviceUrl)
    prop.setProperty(ADMIN_URL_OPTION_KEY, adminUrl)
    prop.setProperty(PARTITION_DISCOVERY_INTERVAL_MS, "5000")
    prop.setProperty(STARTING_OFFSETS_OPTION_KEY, "earliest")
    prop
  }

  def sinkProperties(): Properties = {
    val prop = new Properties()
    prop.setProperty(SERVICE_URL_OPTION_KEY, serviceUrl)
    prop.setProperty(ADMIN_URL_OPTION_KEY, adminUrl)
    prop.setProperty(FLUSH_ON_CHECKPOINT, "true")
    prop.setProperty(FAIL_ON_WRITE, "true")
    prop
  }

  private val topicId = new AtomicInteger(0)

  private def newTopic(): String = TopicName.get(s"topic-${topicId.getAndIncrement()}").toString
}

File: src/test/java/org/apache/flink/streaming/connectors/pulsar/SchemaITest.java
/**
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.streaming.connectors.pulsar;

import org.apache.flink.api.common.typeinfo.TypeInformation;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.connectors.pulsar.internal.PulsarOptions;
import org.apache.flink.streaming.connectors.pulsar.testutils.FailingIdentityMapper;
import org.apache.flink.streaming.connectors.pulsar.testutils.SingletonStreamSink;
import org.apache.flink.table.api.DataTypes;
import org.apache.flink.table.api.Table;
import org.apache.flink.table.api.TableSchema;
import org.apache.flink.table.api.java.StreamTableEnvironment;
import org.apache.flink.table.descriptors.ConnectorDescriptor;
import org.apache.flink.table.descriptors.Pulsar;
import org.apache.flink.table.descriptors.Schema;
import org.apache.flink.table.types.DataType;
import org.apache.flink.table.types.utils.LegacyTypeInfoDataTypeConverter;
import org.apache.flink.util.StringUtils;
import org.apache.pulsar.common.naming.TopicName;
import org.apache.pulsar.common.schema.SchemaType;
import org.junit.Before;
import org.junit.Test;

import java.sql.Timestamp;
import java.text.SimpleDateFormat;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.function.Function;
import java.util.stream.Collectors;

import static org.apache.flink.streaming.connectors.pulsar.SchemaData.booleanList;
import static org.apache.flink.streaming.connectors.pulsar.SchemaData.bytesList;
import static org.apache.flink.streaming.connectors.pulsar.SchemaData.dateList;
import static org.apache.flink.streaming.connectors.pulsar.SchemaData.doubleList;
import static org.apache.flink.streaming.connectors.pulsar.SchemaData.floatList;
import static org.apache.flink.streaming.connectors.pulsar.SchemaData.int16List;
import static org.apache.flink.streaming.connectors.pulsar.SchemaData.int32List;
import static org.apache.flink.streaming.connectors.pulsar.SchemaData.int64List;
import static org.apache.flink.streaming.connectors.pulsar.SchemaData.int8List;
import static org.apache.flink.streaming.connectors.pulsar.SchemaData.stringList;
import static org.apache.flink.streaming.connectors.pulsar.SchemaData.timestampList;

public class SchemaITest extends PulsarTestBaseWithFlink {

    @Before
    public void clearState() {
        SingletonStreamSink.clear();
        FailingIdentityMapper.failedBefore = false;
    }

    @Test
    public void testBooleanRead() throws Exception {
        checkRead(SchemaType.BOOLEAN, booleanList, null, null);
    }

    @Test
    public void testBooleanWrite() throws Exception {
        checkWrite(SchemaType.BOOLEAN, DataTypes.BOOLEAN(), booleanList, null, null);
    }

    @Test
    public void testINT32Read() throws Exception {
        checkRead(SchemaType.INT32, int32List, null, null);
    }

    @Test
    public void testINT32Write() throws Exception {
        checkWrite(SchemaType.INT32, DataTypes.INT(), int32List, null, null);
    }

    @Test
    public void testINT64Read() throws Exception {
        checkRead(SchemaType.INT64, int64List, null, null);
    }

    @Test
    public void testINT64Write() throws Exception {
        checkWrite(SchemaType.INT64, DataTypes.BIGINT(), int64List, null, null);
    }


    @Test
    public void testStringRead() throws Exception {
        checkRead(SchemaType.STRING, stringList, null, null);
    }

    @Test
    public void testStringWrite() throws Exception {
        checkWrite(SchemaType.STRING, DataTypes.STRING(), stringList, null, null);
    }


    @Test
    public void testByteRead() throws Exception {
        checkRead(SchemaType.INT8, int8List, null, null);
    }

    @Test
    public void testByteWrite() throws Exception {
        checkWrite(SchemaType.INT8, DataTypes.TINYINT(), int8List, null, null);
    }


    @Test
    public void testShortRead() throws Exception {
        checkRead(SchemaType.INT16, int16List, null, null);
    }

    @Test
    public void testShortWrite() throws Exception {
        checkWrite(SchemaType.INT16, DataTypes.SMALLINT(), int16List, null, null);
    }


    @Test
    public void testFloatRead() throws Exception {
        checkRead(SchemaType.FLOAT, floatList, null, null);
    }

    @Test
    public void testFloatWrite() throws Exception {
        checkWrite(SchemaType.FLOAT, DataTypes.FLOAT(), floatList, null, null);
    }


    @Test
    public void testDoubleRead() throws Exception {
        checkRead(SchemaType.DOUBLE, doubleList, null, null);
    }

    @Test
    public void testDoubleWrite() throws Exception {
        checkWrite(SchemaType.DOUBLE, DataTypes.DOUBLE(), doubleList, null, null);
    }

    @Test
    public void testDateRead() throws Exception {
        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd");
        checkRead(SchemaType.DATE, dateList, t -> dateFormat.format(t), null);
    }

    @Test
    public void testDateWrite() throws Exception {
        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd");
        checkWrite(SchemaType.DATE,
                DataTypes.DATE().bridgedTo(java.sql.Date.class),
                dateList.stream().map(t -> new java.sql.Date(t.getTime())).collect(Collectors.toList()),
                t -> dateFormat.format(t), null);
    }

    @Test
    public void testTimestampRead() throws Exception {
        checkRead(SchemaType.TIMESTAMP, timestampList, null, null);
    }

    @Test
    public void testTimestampWrite() throws Exception {
        checkWrite(SchemaType.TIMESTAMP,
                DataTypes.TIMESTAMP(3).bridgedTo(Timestamp.class), timestampList, null, null);
    }

    @Test
    public void testByteArrayRead() throws Exception {
        checkRead(SchemaType.BYTES, bytesList, t -> StringUtils.arrayAwareToString(t), null);
    }

    @Test
    public void testByteArrayWrite() throws Exception {
        checkWrite(SchemaType.BYTES, DataTypes.BYTES(), bytesList, t -> StringUtils.arrayAwareToString(t), null);
    }

    private <T> void checkRead(SchemaType type, List<T> datas, Function<T, String> toStr, Class<T> tClass) throws Exception {
        StreamExecutionEnvironment see = StreamExecutionEnvironment.getExecutionEnvironment();
        see.setParallelism(1);
        StreamTableEnvironment tEnv = StreamTableEnvironment.create(see);

        String table = newTopic();
        sendTypedMessages(table, type, datas, Optional.empty(), tClass);

        tEnv.connect(getPulsarSourceDescriptor(table))
                .inAppendMode()
                .registerTableSource(table);

        Table t = tEnv.scan(table).select("value");
        tEnv.toAppendStream(t, t.getSchema().toRowType())
                .map(new FailingIdentityMapper<>(datas.size()))
                .addSink(new SingletonStreamSink.StringSink<>()).setParallelism(1);

        Thread reader = new Thread("read") {
            @Override
            public void run() {
                try {
                    see.execute("read from earliest");
                } catch (Throwable e) {
                    // do nothing
                }
            }
        };

        reader.start();
        reader.join();

        if (toStr == null) {
            SingletonStreamSink.compareWithList(datas.subList(0, datas.size() - 1).stream().map(Objects::toString).collect(Collectors.toList()));
        } else {
            SingletonStreamSink.compareWithList(datas.subList(0, datas.size() - 1).stream().map(e -> toStr.apply(e)).collect(Collectors.toList()));
        }
    }

    private <T> void checkWrite(SchemaType type, DataType dt, List<T> datas, Function<T, String> toStr, Class<T> tClass) throws Exception {
        StreamExecutionEnvironment see = StreamExecutionEnvironment.getExecutionEnvironment();
        see.setParallelism(1);
        StreamTableEnvironment tEnv = StreamTableEnvironment.create(see);

        String table = newTopic();
        String tableName = TopicName.get(table).getLocalName();

        TableSchema tSchema = TableSchema.builder().field("value", dt).build();

        TypeInformation<T> ti = (TypeInformation<T>) LegacyTypeInfoDataTypeConverter.toLegacyTypeInfo(dt);
        DataStream stream = see.fromCollection(datas, ti);
        tEnv.registerDataStream("origin", stream);

        tEnv.connect(getPulsarSinkDescriptor(table))
                .withSchema(new Schema().schema(tSchema))
                .inAppendMode()
                .registerTableSink(tableName);

        tEnv.sqlUpdate("insert into `" + tableName + "` select * from origin");

        Thread sinkThread = new Thread("sink") {
            @Override
            public void run() {
                try {
                    see.execute("sink");
                } catch (Throwable e) {
                    // do nothing
                }
            }
        };

        sinkThread.start();
        sinkThread.join();

        StreamExecutionEnvironment se2 = StreamExecutionEnvironment.getExecutionEnvironment();
        se2.setParallelism(1);
        StreamTableEnvironment tEnv2 = StreamTableEnvironment.create(se2);

        tEnv2.connect(getPulsarSourceDescriptor(table))
                .inAppendMode()
                .registerTableSource(table);


        Table t = tEnv2.scan(table).select("value");
        tEnv2.toAppendStream(t, t.getSchema().toRowType())
                .map(new FailingIdentityMapper<>(datas.size()))
                .addSink(new SingletonStreamSink.StringSink<>()).setParallelism(1);

        FailingIdentityMapper.failedBefore = false;
        SingletonStreamSink.clear();

        Thread reader = new Thread("read") {
            @Override
            public void run() {
                try {
                    se2.execute("read");
                } catch (Throwable e) {
                    // do nothing
                }
            }
        };

        reader.start();
        reader.join();

        if (toStr == null) {
            SingletonStreamSink.compareWithList(datas.subList(0, datas.size() - 1).stream().map(Objects::toString).collect(Collectors.toList()));
        } else {
            SingletonStreamSink.compareWithList(datas.subList(0, datas.size() - 1).stream().map(e -> toStr.apply(e)).collect(Collectors.toList()));
        }
    }


    private ConnectorDescriptor getPulsarSourceDescriptor(String tableName) {
        return new Pulsar()
                .urls(getServiceUrl(), getAdminUrl())
                .topic(tableName)
                .startFromEarliest()
                .property(PulsarOptions.PARTITION_DISCOVERY_INTERVAL_MS_OPTION_KEY, "5000");
    }

    private ConnectorDescriptor getPulsarSinkDescriptor(String tableName) {
        return new Pulsar()
                .urls(getServiceUrl(), getAdminUrl())
                .topic(tableName)
                .property(PulsarOptions.FLUSH_ON_CHECKPOINT_OPTION_KEY, "true")
                .property(PulsarOptions.FAIL_ON_WRITE_OPTION_KEY, "true");
    }
}
